File: python/src/hron/_display.py
from __future__ import annotations

from ._ast import (
    DayFilter,
    DayFilterDays,
    DayFilterEvery,
    DayFilterWeekday,
    DayFilterWeekend,
    DayRange,
    DayRepeat,
    DaysTarget,
    IntervalRepeat,
    IntervalUnit,
    IsoDate,
    IsoException,
    IsoUntil,
    LastDayTarget,
    LastWeekdayTarget,
    MonthRepeat,
    NamedDate,
    NamedException,
    NamedUntil,
    NearestDirection,
    NearestWeekdayTarget,
    OrdinalRepeat,
    ScheduleData,
    ScheduleExpr,
    SingleDateExpr,
    SingleDay,
    TimeOfDay,
    WeekRepeat,
    YearDateTarget,
    YearDayOfMonthTarget,
    YearLastWeekdayTarget,
    YearOrdinalWeekdayTarget,
    YearRepeat,
)


def display(schedule: ScheduleData) -> str:
    out = _display_expr(schedule.expr)

    if schedule.except_:
        parts: list[str] = []
        for exc in schedule.except_:
            match exc:
                case NamedException(month=m, day=d):
                    parts.append(f"{m} {d}")
                case IsoException(date=d):
                    parts.append(d)
        out += " except " + ", ".join(parts)

    if schedule.until:
        match schedule.until:
            case IsoUntil(date=d):
                out += f" until {d}"
            case NamedUntil(month=m, day=d):
                out += f" until {m} {d}"

    if schedule.anchor:
        out += f" starting {schedule.anchor}"

    if schedule.during:
        out += " during " + ", ".join(str(m) for m in schedule.during)

    if schedule.timezone:
        out += f" in {schedule.timezone}"

    return out


def _display_expr(expr: ScheduleExpr) -> str:
    match expr:
        case IntervalRepeat(interval=interval, unit=unit, from_time=ft, to_time=tt, day_filter=df):
            out = f"every {interval} {_unit_display(interval, unit)}"
            out += f" from {ft} to {tt}"
            if df is not None:
                out += f" on {_display_day_filter(df)}"
            return out

        case DayRepeat(interval=interval, days=days, times=times):
            if interval > 1:
                return f"every {interval} days at {_format_time_list(times)}"
            return f"every {_display_day_filter(days)} at {_format_time_list(times)}"

        case WeekRepeat(interval=interval, days=days, times=times):
            day_str = ", ".join(str(d) for d in days)
            return f"every {interval} weeks on {day_str} at {_format_time_list(times)}"

        case MonthRepeat(interval=interval, target=target, times=times):
            match target:
                case DaysTarget(specs=specs):
                    target_str = _format_ordinal_day_specs(specs)
                case LastDayTarget():
                    target_str = "last day"
                case LastWeekdayTarget():
                    target_str = "last weekday"
                case NearestWeekdayTarget(day=day, direction=direction):
                    prefix = ""
                    if direction == NearestDirection.NEXT:
                        prefix = "next "
                    elif direction == NearestDirection.PREVIOUS:
                        prefix = "previous "
                    target_str = f"{prefix}nearest weekday to {day}{_ordinal_suffix(day)}"
                case _:
                    target_str = "last weekday"
            if interval > 1:
                return f"every {interval} months on the {target_str} at {_format_time_list(times)}"
            return f"every month on the {target_str} at {_format_time_list(times)}"

        case OrdinalRepeat(interval=interval, ordinal=ordinal, day=day, times=times):
            if interval > 1:
                return f"{ordinal} {day} of every {interval} months at {_format_time_list(times)}"
            return f"{ordinal} {day} of every month at {_format_time_list(times)}"

        case SingleDateExpr(date=date_spec, times=times):
            match date_spec:
                case NamedDate(month=m, day=d):
                    date_str = f"{m} {d}"
                case IsoDate(date=d):
                    date_str = d
            return f"on {date_str} at {_format_time_list(times)}"

        case YearRepeat(interval=interval, target=target, times=times):
            match target:
                case YearDateTarget(month=m, day=d):
                    target_str = f"{m} {d}"
                case YearOrdinalWeekdayTarget(ordinal=o, weekday=w, month=m):
                    target_str = f"the {o} {w} of {m}"
                case YearDayOfMonthTarget(day=d, month=m):
                    target_str = f"the {d}{_ordinal_suffix(d)} of {m}"
                case YearLastWeekdayTarget(month=m):
                    target_str = f"the last weekday of {m}"
            if interval > 1:
                return f"every {interval} years on {target_str} at {_format_time_list(times)}"
            return f"every year on {target_str} at {_format_time_list(times)}"

    # Should be unreachable
    raise ValueError(f"unknown expression type: {type(expr)}")  # pragma: no cover


def _display_day_filter(f: DayFilter) -> str:
    match f:
        case DayFilterEvery():
            return "day"
        case DayFilterWeekday():
            return "weekday"
        case DayFilterWeekend():
            return "weekend"
        case DayFilterDays(days=days):
            return ", ".join(str(d) for d in days)


def _format_time_list(times: tuple[TimeOfDay, ...]) -> str:
    return ", ".join(str(t) for t in times)


def _format_ordinal_day_specs(specs: tuple[SingleDay | DayRange, ...]) -> str:
    parts: list[str] = []
    for spec in specs:
        match spec:
            case SingleDay(day=d):
                parts.append(f"{d}{_ordinal_suffix(d)}")
            case DayRange(start=s, end=e):
                parts.append(f"{s}{_ordinal_suffix(s)} to {e}{_ordinal_suffix(e)}")
    return ", ".join(parts)


def _ordinal_suffix(n: int) -> str:
    mod100 = n % 100
    if 11 <= mod100 <= 13:
        return "th"
    match n % 10:
        case 1:
            return "st"
        case 2:
            return "nd"
        case 3:
            return "rd"
        case _:
            return "th"


def _unit_display(interval: int, unit: IntervalUnit) -> str:
    if unit == IntervalUnit.MIN:
        return "minute" if interval == 1 else "min"
    return "hour" if interval == 1 else "hours"

File: ruby/lib/hron/display.rb
# frozen_string_literal: true

require_relative "ast"

module Hron
  # Display module for converting schedules to canonical string representation
  module Display
    def self.display(schedule)
      out = display_expr(schedule.expr)

      unless schedule.except.empty?
        parts = schedule.except.map do |exc|
          case exc
          when NamedException
            "#{exc.month} #{exc.day}"
          when IsoException
            exc.date
          else
            raise "unknown exception type: #{exc.class}"
          end
        end
        out += " except #{parts.join(", ")}"
      end

      if schedule.until
        case schedule.until
        when IsoUntil
          out += " until #{schedule.until.date}"
        when NamedUntil
          out += " until #{schedule.until.month} #{schedule.until.day}"
        else
          raise "unknown until type: #{schedule.until.class}"
        end
      end

      out += " starting #{schedule.anchor}" if schedule.anchor

      out += " during #{schedule.during.join(", ")}" unless schedule.during.empty?

      out += " in #{schedule.timezone}" if schedule.timezone

      out
    end

    def self.display_expr(expr)
      case expr
      when IntervalRepeat
        out = "every #{expr.interval} #{unit_display(expr.interval, expr.unit)}"
        out += " from #{expr.from_time} to #{expr.to_time}"
        out += " on #{display_day_filter(expr.day_filter)}" if expr.day_filter
        out

      when DayRepeat
        if expr.interval > 1
          "every #{expr.interval} days at #{format_time_list(expr.times)}"
        else
          "every #{display_day_filter(expr.days)} at #{format_time_list(expr.times)}"
        end

      when WeekRepeat
        day_str = expr.days.join(", ")
        "every #{expr.interval} weeks on #{day_str} at #{format_time_list(expr.times)}"

      when MonthRepeat
        target_str = case expr.target
        when DaysTarget
          format_ordinal_day_specs(expr.target.specs)
        when LastDayTarget
          "last day"
        when LastWeekdayTarget
          "last weekday"
        when NearestWeekdayTarget
          prefix = case expr.target.direction
          when NearestDirection::NEXT
            "next "
          when NearestDirection::PREVIOUS
            "previous "
          else
            ""
          end
          day = expr.target.day
          "#{prefix}nearest weekday to #{day}#{ordinal_suffix(day)}"
        when OrdinalWeekdayTarget
          "#{expr.target.ordinal} #{expr.target.weekday}"
        else
          raise "unknown month target: #{expr.target.class}"
        end
        if expr.interval > 1
          "every #{expr.interval} months on the #{target_str} at #{format_time_list(expr.times)}"
        else
          "every month on the #{target_str} at #{format_time_list(expr.times)}"
        end

      when SingleDateExpr
        date_str = case expr.date
        when NamedDate
          "#{expr.date.month} #{expr.date.day}"
        when IsoDate
          expr.date.date
        else
          raise "unknown date type: #{expr.date.class}"
        end
        "on #{date_str} at #{format_time_list(expr.times)}"

      when YearRepeat
        target_str = case expr.target
        when YearDateTarget
          "#{expr.target.month} #{expr.target.day}"
        when YearOrdinalWeekdayTarget
          "the #{expr.target.ordinal} #{expr.target.weekday} of #{expr.target.month}"
        when YearDayOfMonthTarget
          "the #{expr.target.day}#{ordinal_suffix(expr.target.day)} of #{expr.target.month}"
        when YearLastWeekdayTarget
          "the last weekday of #{expr.target.month}"
        else
          raise "unknown year target: #{expr.target.class}"
        end
        if expr.interval > 1
          "every #{expr.interval} years on #{target_str} at #{format_time_list(expr.times)}"
        else
          "every year on #{target_str} at #{format_time_list(expr.times)}"
        end

      else
        raise "unknown expression type: #{expr.class}"
      end
    end

    def self.display_day_filter(filter)
      case filter
      when DayFilterEvery
        "day"
      when DayFilterWeekday
        "weekday"
      when DayFilterWeekend
        "weekend"
      when DayFilterDays
        filter.days.join(", ")
      else
        raise "unknown day filter: #{filter.class}"
      end
    end

    def self.format_time_list(times)
      times.map(&:to_s).join(", ")
    end

    def self.format_ordinal_day_specs(specs)
      parts = specs.map do |spec|
        case spec
        when SingleDay
          "#{spec.day}#{ordinal_suffix(spec.day)}"
        when DayRange
          "#{spec.start}#{ordinal_suffix(spec.start)} to #{spec.end_day}#{ordinal_suffix(spec.end_day)}"
        end
      end
      parts.join(", ")
    end

    def self.ordinal_suffix(n)
      mod100 = n % 100
      return "th" if mod100.between?(11, 13)

      case n % 10
      when 1
        "st"
      when 2
        "nd"
      when 3
        "rd"
      else
        "th"
      end
    end

    def self.unit_display(interval, unit)
      if unit == IntervalUnit::MIN
        (interval == 1) ? "minute" : "min"
      else
        (interval == 1) ? "hour" : "hours"
      end
    end
  end
end
